File: problems/07-monotonic-stack/002-next-greater-element/python/solution.py
class Solution:
    def nextGreaterElements(self, nums: list[int]) -> list[int]:
        n = len(nums)
        answer = [-1] * n
        stack = []

        # Traverse twice to simulate a circular array.
        for i in range(2 * n - 1, -1, -1):
            index = i % n

            while stack and stack[-1] <= nums[index]:
                stack.pop()

            if i < n and stack:
                answer[index] = stack[-1]

            stack.append(nums[index])

        return answer

File: problems/07-monotonic-stack/002-next-greater-element/java/Solution.java
import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.Deque;

public class Solution {
    public int[] nextGreaterElements(int[] nums) {
        int n = nums.length;
        int[] answer = new int[n];
        Arrays.fill(answer, -1);
        Deque<Integer> stack = new ArrayDeque<>();

        // Traverse twice to simulate a circular array.
        for (int i = 2 * n - 1; i >= 0; i--) {
            int index = i % n;

            while (!stack.isEmpty() && stack.peek() <= nums[index]) {
                stack.pop();
            }

            if (i < n && !stack.isEmpty()) {
                answer[index] = stack.peek();
            }

            stack.push(nums[index]);
        }

        return answer;
    }
}
